rust/otap-dataflow/crates/otap/src/grpc.rs

```diff
@@ -0,0 +1,235 @@
+//! ToDo: Modify OTAPData -> Optimize message transport
+//! ToDo: Handle Ack and Nack, return proper batch status
+//! ToDo: Change how channel sizes are handled? Currently defined when creating otap_receiver -> passing channel size to the ServiceImpl
+//!
+
+use crate::proto::opentelemetry::experimental::arrow::v1::{
+    BatchArrowRecords, BatchStatus, StatusCode, arrow_logs_service_server::ArrowLogsService,
+    arrow_metrics_service_server::ArrowMetricsService,
+    arrow_traces_service_server::ArrowTracesService,
+};
+use otap_df_engine::error::Error;
+use otap_df_engine::shared::receiver as shared;
+use std::pin::Pin;
+use tokio_stream::Stream;
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::codec::CompressionEncoding;
+use tonic::{Request, Response, Status};
+
+/// struct that implements the ArrowLogsService trait
+pub struct ArrowLogsServiceImpl {
+    effect_handler: shared::EffectHandler<OTAPData>,
+    message_size: usize,
+}
+
+impl ArrowLogsServiceImpl {
+    /// create a new ArrowLogsServiceImpl struct with a sendable effect handler
+    #[must_use]
+    pub fn new(effect_handler: shared::EffectHandler<OTAPData>, message_size: usize) -> Self {
+        Self {
+            effect_handler,
+            message_size,
+        }
+    }
+}
+/// struct that implements the ArrowMetricsService trait
+pub struct ArrowMetricsServiceImpl {
+    effect_handler: shared::EffectHandler<OTAPData>,
+    message_size: usize,
+}
+
+impl ArrowMetricsServiceImpl {
+    /// create a new ArrowMetricsServiceImpl struct with a sendable effect handler
+    #[must_use]
+    pub fn new(effect_handler: shared::EffectHandler<OTAPData>, message_size: usize) -> Self {
+        Self {
+            effect_handler,
+            message_size,
+        }
+    }
+}
+
+/// struct that implements the ArrowTracesService trait
+pub struct ArrowTracesServiceImpl {
+    effect_handler: shared::EffectHandler<OTAPData>,
+    message_size: usize,
+}
+
+impl ArrowTracesServiceImpl {
+    /// create a new ArrowTracesServiceImpl struct with a sendable effect handler
+    #[must_use]
+    pub fn new(effect_handler: shared::EffectHandler<OTAPData>, message_size: usize) -> Self {
+        Self {
+            effect_handler,
+            message_size,
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl ArrowLogsService for ArrowLogsServiceImpl {
+    type ArrowLogsStream =
+        Pin<Box<dyn Stream<Item = Result<BatchStatus, Status>> + Send + 'static>>;
+    async fn arrow_logs(
+        &self,
+        request: Request<tonic::Streaming<BatchArrowRecords>>,
+    ) -> Result<Response<Self::ArrowLogsStream>, Status> {
+        let mut input_stream = request.into_inner();
+        let (tx, rx) = tokio::sync::mpsc::channel(self.message_size);
+        let effect_handler_clone = self.effect_handler.clone();
+
+        // Provide client a stream to listen to
+        let output = ReceiverStream::new(rx);
+
+        // write to the channel
+        _ = tokio::spawn(async move {
```

Comment:
```suggestion
        // ToDo [LQ] How can we abstract this to avoid any dependency on Tokio inside receiver implementations.
        _ = tokio::spawn(async move {
```